File: py/sort/merge_sort.py
def merge_sort(arr):
    # base case: array with 0 or 1 element is already sorted
    if len(arr) <= 1:
        return arr

    # divide the array into two halves
    mid = len(arr) // 2
    left_half = arr[:mid]
    right_half = arr[mid:]

    # recursively sort each half
    left_half = merge_sort(left_half)
    right_half = merge_sort(right_half)

    # merge the sorted halves
    sorted_arr = []
    i, j = 0, 0
    while i < len(left_half) and j < len(right_half):
        if left_half[i] <= right_half[j]:
            sorted_arr.append(left_half[i])
            i += 1
        else:
            sorted_arr.append(right_half[j])
            j += 1
    sorted_arr += left_half[i:]
    sorted_arr += right_half[j:]

    return sorted_arr

File: js/sort/merge_sort.js
function mergeSort(arr) {
  if (arr.length < 2) {
    return arr;
  }

  const middle = Math.floor(arr.length / 2);
  const left = arr.slice(0, middle);
  const right = arr.slice(middle);

  return merge(mergeSort(left), mergeSort(right));
}

function merge(left, right) {
  let result = [];

  while (left.length && right.length) {
    if (left[0] <= right[0]) {
      result.push(left.shift());
    } else {
      result.push(right.shift());
    }
  }

  while (left.length) {
    result.push(left.shift());
  }

  while (right.length) {
    result.push(right.shift());
  }

  return result;
}

const arr = [5, 2, 4, 7, 1, 3, 2, 6];
const sortedArr = mergeSort(arr);
console.log(sortedArr); // [1, 2, 2, 3, 4, 5, 6, 7]

/*
This implementation uses a divide-and-conquer approach to recursively split the input array in half until each 
element is in its own array. Then, it merges the smaller arrays back together in sorted order using the merge 
function. The merge function takes two sorted arrays and combines them into a single sorted array.

This implementation has a time complexity of O(n log n) and a space complexity of O(n).
*/